### PR TITLE
ENH: Allow DataBroker['(parital) uid string']

### LINKED ISSUE #130: Teach DataBroker[key] about uids
The `DataBroker[key]` syntax supports only slices or `scan_id` integers which may not be unique.  Please, let it also digest the `run_start_uid`, `run_stop_uid` values of the document object, and ideally their partials strings.

``` python
from dataportal.broker import DataBroker as db
header = db[-1]
assert header == db[header.run_start_uid]
assert header == db[header.run_start_uid[:7]]
```

### PR DESCRIPTION
**Pull request by danielballan**
As suggested by @pavoljuhas, you can search for `<Header scan_id=1 run_start_uid='d4b78e8b-a23d-4d79-8c9a-3531a86a84d8'>` by its uid string or the first few characters of that string.

```
In [3]: DataBroker['d4b78e']
Out[3]: <Header scan_id=1 run_start_uid='d4b78e8b-a23d-4d79-8c9a-3531a86a84d8'>
```

If a given partial string is not specific enough, it raises informatively.

```
In [4]: DataBroker['e']
ValueError: That partial uid matches multiple runs. Provide more characters.
```

Closes #130

### CODE REVIEW

File: dataportal/broker/simple_broker.py [hunk 1]
```diff
@@ -58,12 +60,31 @@ def __getitem__(cls, key):
                         raise IndexError(
                             "There are only {0} runs.".format(i))
                 header = Header.from_run_start(result)
+        elif isinstance(key, six.string_types):
+            # Interpret key as a uid (or the few several characters of one).
+            if len(key) == 36:
```
Comment (tacaswell) [inline dataportal/broker/simple_broker.py]: Not a fan of hard-coding this.  Is the regex a big performance hit?

Comment (tacaswell) [inline dataportal/broker/simple_broker.py]: Maybe try the query with `uid=key` and then if that returns nothing, try with the regex?

Comment (danielballan) [inline dataportal/broker/simple_broker.py]: Clever. New commit does this.

**Comment by tacaswell** [comment]
Closed by #359

**Pull request merged**